nengo_spa/modules/assoc_mem.py

```diff
@@ -54,27 +58,28 @@ class AssociativeMemory(Network):
         'output_vocab', default=None, readonly=True)
 
     def __init__(
-            self, selection_net, input_vocab, output_vocab=None,
-            input_keys=None, output_keys=None, n_neurons=50,
-            label=None, seed=None, add_to_container=None, vocabs=None,
-            **selection_net_args):
+            self, selection_net, input_vocab, output_vocab=None, mapping=None,
+            n_neurons=50, label=None, seed=None, add_to_container=None,
+            vocabs=None, **selection_net_args):
         super(AssociativeMemory, self).__init__(
             label=label, seed=seed, add_to_container=add_to_container,
             vocabs=vocabs)
 
         if output_vocab is None:
             output_vocab = input_vocab
+        elif mapping is None:
+            raise ValidationError(
+                "The mapping argument needs to be provided if an output "
+                "vocabulary is given.", attr='mapping', obj=self)
         self.input_vocab = input_vocab
         self.output_vocab = output_vocab
 
-        if input_keys is None:
-            input_keys = self.input_vocab.keys()
-            input_vectors = self.input_vocab.vectors
-        else:
-            input_vectors = [input_vocab.parse(key).v for key in input_keys]
+        if mapping is None or mapping == 'by-key':
```

Comment:
In the string case that might make sense, but I probably would not check the type. That is always a bit problematic with duck typing because you don't really want to check for the type, but whether the object behaves the right way (in this case it needs to have a `keys` method returning an iterable and support indexing with the elements).